{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "------------------------------------------------------------------------\n",
    "\n",
    "Copyright 2023 Christopher Fazekas [Alejandro Chavez Lab at UCSD]\n",
    "\n",
    "All Rights Reserved\n",
    "\n",
    "MAVDA Academic License\n",
    "\n",
    "preprocess.ipynb\n",
    "\n",
    "------------------------------------------------------------------------"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SeqIO.QualityIO import FastqGeneralIterator\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os\n",
    "import glob\n",
    "import gzip\n",
    "pd.options.mode.chained_assignment = None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "constants - 1. a merged.csv file, containing indexing information for all possible primer plates\n",
    "\n",
    "            ,Well,7-series_ID,5-series_ID,7-series_SEQ,5-series_SEQ,Internal_ID,Internal_Seq,PP,SAMPLE_ID,merged_index\n",
    "            0,A1,n701,s502,TAAGGCGA,ATAGAGAG,1,AATGC,1,A170,TAAGGCGAATAGAGAGAATGC\n",
    "            1,A2,n701,s502,TAAGGCGA,ATAGAGAG,2,GTTGC,1,A170,TAAGGCGAATAGAGAGGTTGC\n",
    "            2,A3,n701,s502,TAAGGCGA,ATAGAGAG,3,GATCA,1,A170,TAAGGCGAATAGAGAGGATCA\n",
    "            3,A4,n701,s502,TAAGGCGA,ATAGAGAG,4,TGATC,1,A170,TAAGGCGAATAGAGAGTGATC\n",
    "            4,A5,n701,s502,TAAGGCGA,ATAGAGAG,5,ACTCC,1,A170,TAAGGCGAATAGAGAGACTCC\n",
    "            5,A6,n701,s502,TAAGGCGA,ATAGAGAG,6,ACCTG,1,A170,TAAGGCGAATAGAGAGACCTG   \n",
    "\n",
    "\n",
    "input - 1. database of fastq.gz files, already aligned to illuminia indicies\n",
    "        get indexes (i7,i5) as well as internal barcode index and model index\n",
    "        which are both at a fixed position in the read \n",
    "\n",
    "        2. A setup csv file, containing the plate ID (names) and associated primer \n",
    "        plate. \n",
    "\n",
    "        plateID,primerplate\n",
    "        NS1519_22057,1\n",
    "        NS1519_22058,13\n",
    "        NS1519_22059,7\n",
    "        NS1519_22060,10\n",
    "\n",
    "        3. tbd format of OD file (??? not required for counts, maybe for backend)\n",
    "\n",
    "output - \n",
    "        1. counts table of the format:\n",
    "\n",
    "                    wellid, wellid, wellid, wellid, wellid  x96\n",
    "            model       25      7       29      30      20\n",
    "            model       23      2       23      15      10\n",
    "            model       12      4       12      35      15\n",
    "            x217\n",
    "\n",
    "for each platename and primer plate number in the setup csv, pull corresponding .gz files into \n",
    "a separate working folder. \n",
    "\n",
    "generate a flat dictionary (countsdict) where {key:value} == {wellid_model:count}\n",
    "i.e {GGACTCCTAGAGGATACGATA:1}\n",
    "\n",
    "for each gz in the working folder\n",
    "    set i7 and i5 indicies\n",
    "    skip to 2nd line, get column id and protease id\n",
    "    set wellid_model \n",
    "    countsdict[wellid_model] += 1\n",
    "\n",
    "convert the dictionary to an output plateid_counts.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "setupname = \"setup.csv\"\n",
    "\n",
    "\n",
    "\n",
    "direc = os.getcwd()\n",
    "setuppath = direc + \"/\" + setupname\n",
    "mergedpath = direc + \"/\" + \"merged.csv\"\n",
    "models = direc + \"/\" + \"models.csv\"\n",
    "\n",
    "#reading in runinfo - plateIDs and assocaited primer plate IDs \n",
    "r = pd.read_csv(setuppath)\n",
    "runinfo = r.set_index(\"plateID\").to_dict()\n",
    "\n",
    "#reading in merged dataframe\n",
    "m = pd.read_csv(mergedpath, index_col=0)\n",
    "\n",
    "#reading in models\n",
    "g = pd.read_csv(models, index_col=0)\n",
    "\n",
    "def returnrunindices(merged,primerplatenum):\n",
    "    #takes merged primer plate frame and primer plate number and returns a dataframe\n",
    "    #containing only the indices from the primer plate\n",
    "    returndf = merged[merged[\"PP\"]==primerplatenum]\n",
    "    return returndf\n",
    "\n",
    "def modelwiseindicies(runindicies, modelcsv):\n",
    "    #takes the well incidies for a plate (or plates) and returns a dataframe containing the model-wise\n",
    "    #index for each well\n",
    "    modellst = list(modelcsv[\"model\"] + \"_\" + modelcsv[\"seq\"])\n",
    "    runindicies[\"model_seq\"] = runindicies.apply(lambda x: modellst, axis=1)\n",
    "    retdf = runindicies.explode(\"model_seq\")\n",
    "    retdf[\"fullseq\"] = retdf[\"merged_index\"].str[:] + retdf[\"model_seq\"].str[-10:]\n",
    "    return retdf\n",
    "\n",
    "print(r)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for plate in runinfo['primerplate'].items():\n",
    "    plateID = plate[0]\n",
    "    ppID = plate[1]\n",
    "    indexsub = m[m[\"PP\"]==ppID] #get subset of indicies for all models all wells in a plate\n",
    "    subind = modelwiseindicies(indexsub, g)\n",
    "    counts = dict((el,0) for el in subind[\"fullseq\"]) #form counts dictionary for each full index sequence\n",
    "    asers = subind[\"SAMPLE_ID\"].unique() #get associated a-series indicies from the primer plate (lst)\n",
    "    for a in asers:\n",
    "        rowsubind = subind[subind[\"SAMPLE_ID\"]==a]\n",
    "        sevenser = rowsubind[\"7-series_SEQ\"].unique()[0] #grab row 7-series seq\n",
    "        fiveser = rowsubind[\"5-series_SEQ\"].unique()[0] #grab row 5-series seq\n",
    "        for parta in glob.iglob(direc + \"/gz/\" + a + \"*R1_001.fastq.gz\"): #iterator for all .fastq.gz files of type a\n",
    "            readstats = [0,0] #unindexed counts, total counts\n",
    "            with gzip.open(parta, \"rt\") as handle: #open .gz\n",
    "                for title, sequence, quality in FastqGeneralIterator(handle): #iterator for fastq files (returns tup(title,sequence,quality))\n",
    "                    countsind = sevenser + fiveser + sequence[4:9] + sequence[45:55] #form index seqeunce\n",
    "                    readstats[1] += 1\n",
    "                    try:\n",
    "                        counts[countsind] += 1 #increment sequence count by 1\n",
    "                    except:\n",
    "                        readstats[0] += 1\n",
    "                        pass\n",
    "                print(parta + \" total counts:\" + str(readstats[1]) + \"; unaligned:\" + str(readstats[0]) + \"; \" + str(100 - round((readstats[0]/readstats[1])*100,2)) + \"%\" +\" alignment\")\n",
    "    subind[\"counts\"] = subind[\"fullseq\"].map(counts) #map counts dictionary to subind dataframe\n",
    "    totable = subind[[\"Well\",\"model_seq\",\"counts\"]] #create output df from subind, reorganize order, names of columns, set barcode and well, ect. \n",
    "    totable[\"runid\"] = plateID\n",
    "    totable[\"gene\"] = totable[\"model_seq\"].str.split(\"_\").str[0]\n",
    "    totable[\"barcode\"] = totable[\"model_seq\"].str.split(\"_\").str[1]\n",
    "    totable = totable.drop([\"model_seq\"], axis=1)\n",
    "    totable = totable.rename(columns={\"Well\":\"well\"})\n",
    "    finorder = [\"runid\",\"well\",\"gene\",\"barcode\",\"counts\"]\n",
    "    totable = totable.reindex(columns=finorder)\n",
    "    os.makedirs(direc + \"/counttables\", exist_ok=True)\n",
    "    totable.to_csv(direc + \"/counttables/\" + plateID + \"_\" + \"PP\" + str(ppID) + \".csv\", index=False) #output counts csv\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
